segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMetadata.java

```diff
@@ -184,6 +185,13 @@ public synchronized long getLength() {
         return new AttributesView();
     }
 
+    @Override
+    public synchronized Map<UUID, Long> getAttributes(BiPredicate<UUID, Long> filter) {
+        return getAttributes().entrySet().stream()
```

Comment:
This doesn't actually remove the problematic getAttributes() method. It's right above, public, and unsafe.

Reply:
It's not removing it. The method above can be used for retrieving values. It needs to be there. It's just not good for iteration.

Reply:
Then perhaps the best way to do that would be to wrap it an make it throw if you attempt to iterate over the values.

Reply:
The `getAttributes` method comes from the `SegmentProperties` interface which just says in it's javadoc "Gets a read-only Map of AttributeId-Values for this Segment.". Nothing would lead you to believe a lock was required. Looking at the code, I can see 10 places where that method is used to return a map that is iterated over, and none of them hold a lock that I can see.

Reply:
The callsites where you see that being iterated over are all covered by a lock. They are only invoked by the OperationProcessor while holding its own lock (same lock used to update them). 

There's a difference between `SegmentProperties` and `SegmentMetadata`. Whomever uses the latter only does so internally, usually inside the pipeline. `SegmentProperties` is used externally and every place that is returned, we create a new object via `SegmentMetadata.getSnapshot` which is thread safe.

Reply:
@tkaitchuck After our private conversation, I added the `getAttributes` in `SegmentMetadata` and explained its behavior in its JavaDoc.